src/hdskins/java/com/voxelmodpack/hdskins/HDSkinManager.java

```diff
@@ -175,124 +169,115 @@ private String bustCache(String url) {
-                    DefaultPlayerSkin.getDefaultSkinLegacy(),
-                    new IImageBuffer() {
-                        @Nonnull
-                        @Override
-                        public BufferedImage parseUserSkin(@Nonnull BufferedImage image) {
-                            BufferedImage image1 = image;
-                            if (imagebufferdownload != null) {
-                                image1 = imagebufferdownload.parseUserSkin(image);
-                            }
-                            return image1 == null ? image : image1;
-                        }
-
-                        @Override
-                        public void skinAvailable() {
-                            if (imagebufferdownload != null) {
-                                imagebufferdownload.skinAvailable();
-                            }
-                            callback.skinAvailable(type, skin, texture);
-                        }
-                    });
-
-            // schedule texture loading on the main thread.
-            TextureLoader.loadTexture(skin, texObject);
+    private void loadTexture(GameProfile profile, Type type, SkinAvailableCallback callback) {
+        if (profile.getId() == null) {
+            return;
         }
+
+        MinecraftProfileTexture texture = getProfileData(profile).get(type);
+
+        String skinDir = type.toString().toLowerCase() + "s/";
+
+        ResourceLocation skin = new ResourceLocation("hdskins", skinDir + texture.getHash());
+
+        File etag = new File(LiteLoader.getAssetsDirectory(), "hd/" + skinDir + texture.getHash().substring(0, 2) + "/" + texture.getHash());
+
+        IImageBuffer buffer = new ImageBufferDownloadHD() {
+            @Nonnull
+            @Override
+            public BufferedImage parseUserSkin(@Nonnull BufferedImage image) {
+                if (type != Type.SKIN) {
+                    return image;
+                }
+
+                BufferedImage converted = super.parseUserSkin(image);
+
+                return converted == null ? image : converted;
+            }
+        }.withCallback(() -> callback.skinAvailable(type, skin, texture));
+
+        // schedule texture loading on the main thread.
+        TextureLoader.loadTexture(skin, new ThreadDownloadImageETag(etag, bustCache(texture.getUrl()), DefaultPlayerSkin.getDefaultSkinLegacy(), buffer));
     }
 
     private Map<Type, MinecraftProfileTexture> loadProfileData(GameProfile profile) {
         Map<Type, MinecraftProfileTexture> textures = Maps.newEnumMap(Type.class);
-        for (SkinServer server : skinServers) {
-            Optional<MinecraftTexturesPayload> profileData = server.loadProfileData(profile);
-            profileData.map(MinecraftTexturesPayload::getTextures).ifPresent(it -> it.forEach(textures::putIfAbsent));
-            if (textures.size() == Type.values().length) {
-                break;
-            }
 
-        }
+        skinServers.forEach(server -> server
```

Comment:
Tried a stream. This easier to write, easier to read, easier to understand, and I'm guessing better for performance. (Since you don't have to construct a stream to construct a stream to join multiple constructed streams into a constructed stream to iterate and construct a stream.)

Reply:
If you really cared about performance, you wouldn't use `forEach`.